File: Problems/Arrays/Product-of-Array-Except-Self/Product-of-Array-Except-Self.cpp
#include <iostream>
#include <vector>

// Using prefix and postfix - Time O(n)

class Solution {
public:
    std::vector<int> productExceptSelf(std::vector<int> &nums)
    {
        int n = nums.size();
        std::vector<int> result(n, 1);
        result[0] = 1;

        for (int i = 1; i < n; i++)
        {
            result[i] = result[i - 1] * nums[i - 1];
        }

        int rightSide = 1;
        for (int i = n - 1; i >= 0; i--)
        {
            result[i] = result[i] * rightSide;
            rightSide *= nums[i]; // rightSide = rightSide * nums[i];
        }

        return result;
    }
};

int main()
{
    std::vector<int> nums = {1, 2, 3, 4};

    // Print input
    std::cout << "Input: nums = [";
    for (const auto &num : nums)
    {
        std::cout << num << "";
        if (&num != &nums.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    Solution sol;
    std::vector<int> result = sol.productExceptSelf(nums);

    // Print input
    std::cout << "Output: [";
    for (const auto &r : result)
    {
        std::cout << r << "";
        if (&r != &result.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    return 0;
}

File: Problems/Arrays/Product-of-Array-Except-Self/Product-of-Array-Except-Self.cs
using System;

// Using prefix and postfix - Time O(n)
public class Solution
{
    public int[] ProductExceptSelf(int[] nums)
    {
        var result = new int[nums.Length];
        result[0] = 1;

        for (int i = 1; i < nums.Length; i++)
        {
            result[i] = result[i - 1] * nums[i - 1];
        }

        int rightSide = 1;
        for (int i = nums.Length - 1; i >= 0; i--)
        {
            result[i] = result[i] * rightSide;
            rightSide *= nums[i]; // rightSide = rightSide * nums[i];
        }

        return result;
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[] nums = { 1, 2, 3, 4 };

        // Print intput
        Console.Write("Input: nums = [");
        foreach (int n in nums)
        {
            Console.Write(n + "");
            if (n != nums[nums.Length - 1])
            {
                Console.Write(", ");
            }
        }
        Console.WriteLine("]");

        Solution sol = new Solution();
        int[] result = sol.ProductExceptSelf(nums);

        // Print intput
        Console.Write("Output: [");
        foreach (int r in result)
        {
            Console.Write(r + "");
            if (r != result[result.Length - 1])
            {
                Console.Write(", ");
            }
        }
        Console.WriteLine("]");
    }
}
